cloudconfig/containerinit/container_userdata_test.go

```diff
@@ -84,80 +88,139 @@ func (s *UserDataSuite) SetUpTest(c *gc.C) {
 		ConfigType:    network.ConfigManual,
 		NoAutoStart:   true,
 	}}
+	s.expectedSampleConfigWriting = `#cloud-config
+bootcmd:
+- install -D -m 644 /dev/null '%[1]s'
+- |-
+  printf '%%s\n' '
+  auto lo {ethaa_bb_cc_dd_ee_f0} {ethaa_bb_cc_dd_ee_f1} {eth}
+
+  iface lo inet loopback
+    dns-nameservers ns1.invalid ns2.invalid
+    dns-search bar foo
+
+  iface {ethaa_bb_cc_dd_ee_f0} inet static
+    address 0.1.2.3/24
+    gateway 0.1.2.1
+
+  iface {ethaa_bb_cc_dd_ee_f1} inet static
+    address 0.2.2.4/24
+    post-up ip route add 0.5.6.0/24 via 0.2.2.1 metric 50
+    pre-down ip route del 0.5.6.0/24 via 0.2.2.1 metric 50
+
+  iface {eth} inet dhcp
+
+  iface {eth} inet dhcp
+
+  iface {eth} inet dhcp
+  ' > '%[1]s'
+`
 	s.expectedSampleConfig = `
-auto any0 any1 any3 lo
+auto lo {ethaa_bb_cc_dd_ee_f0} {ethaa_bb_cc_dd_ee_f1} {eth}
 
 iface lo inet loopback
   dns-nameservers ns1.invalid ns2.invalid
   dns-search bar foo
 
-iface any0 inet static
+iface {ethaa_bb_cc_dd_ee_f0} inet static
   address 0.1.2.3/24
   gateway 0.1.2.1
 
-iface any1 inet static
+iface {ethaa_bb_cc_dd_ee_f1} inet static
   address 0.2.2.4/24
   post-up ip route add 0.5.6.0/24 via 0.2.2.1 metric 50
   pre-down ip route del 0.5.6.0/24 via 0.2.2.1 metric 50
 
-iface any2 inet dhcp
+iface {eth} inet dhcp
 
-iface any3 inet dhcp
+iface {eth} inet dhcp
 
-iface any4 inet manual
+iface {eth} inet dhcp
 `
 	s.expectedSampleUserData = `
-#cloud-config
+- install -D -m 744 /dev/null '%[2]s'
+- |-
+  printf '%%s\n' 'import subprocess, re
+  from string import Formatter
+  INTERFACES_FILE="%[1]s"
+  IP_LINE = re.compile(r"^\d: (.*?):")
+  IP_HWADDR = re.compile(r".*link/ether ((\w{2}|:){11})")
+
+
+  def ip_parse(ip_output):
+      """parses the output of the ip command
+      and returns a hwaddr->nic-name dict"""
+      devices = dict()
+      print ("parsing ip command output")
+      print (ip_output)
+      for ip_line in ip_output:
+          ip_line_str = str(ip_line, '"'"'utf-8'"'"')
+          match = IP_LINE.match(ip_line_str)
+          if match is None:
+              continue
+          nic_name = match.group(1)
+          match = IP_HWADDR.match(ip_line_str)
+          if match is None:
+              continue
+          nic_hwaddr = match.group(1)
+          devices[nic_hwaddr]=nic_name
+      print("found the following devices: " + str(devices))
+      return devices
+
+  def replace_ethernets(interfaces_file, devices):
+      """check if the contents of interfaces_file contain template
+      keys corresponding to hwaddresses and replace them with
+      the proper device name"""
+      interfaces_file_descriptor = open(interfaces_file, "r")
+      interfaces = interfaces_file_descriptor.read()
+      formatter = Formatter()
+      hwaddrs = [v[1] for v in formatter.parse(interfaces) if v[1]]
```

Comment:
are these hardware addresses or are these device names?

it was a little odd to read
  "found the following hwaddrs: [eth10_11_13_14]"

We also might want to put these into a set instead of a list.
We probably also need to be a bit more careful about the parsing, looking for something like
 if v[1] and v[0] == 'iface'